Fix license generator

caneraltinbasak opened the pull request:
License generator is unable to find bulk of licenses in Chromium138. Also formatting of the python file wasn't good. I formatted the file with black first, then added a comprehensive mode which finds all licenses successfully. 

I am planning to rebase these changes on top of master.

caneraltinbasak committed:
generate_license_list: Format with black

Formatted the document with black.

caneraltinbasak committed:
fix(generate_license_list): Do a more comprehensive check on licenses.

Previous version had the problem of giving up if there is a problem
on licenses in one directory. It wasn't recursively checking the inner
directories.

t-bashir-bs commented on meta-chromium/recipes-browser/chromium/generate_license_list.py:
```diff
@@ -83,41 +86,172 @@ def find_chromium_licenses(chromium_root):
-if __name__ == '__main__':
+def _try_find_common_license_files(directory, chromium_root, license_files_set):
+    """Try to find common license file names in a directory when metadata parsing fails."""
+    common_license_names = [
+        "LICENSE",
+        "LICENSE.txt",
+        "LICENSE.md",
+        "LICENSE.rst",
+        "COPYING",
+        "COPYING.txt",
+        "COPYING.md",
+        "NOTICE",
+        "NOTICE.txt",
+        "NOTICE.md",
+        "COPYRIGHT",
+        "COPYRIGHT.txt",
+        "COPYRIGHT.md",
+    ]
```
I agree, it would be good to not duplicate it.

t-bashir-bs commented on meta-chromium/recipes-browser/chromium/generate_license_list.py:
```diff
@@ -86,22 +91,145 @@ def find_chromium_licenses(chromium_root):
+                except (OSError, PermissionError):
+                    pass
+    except (OSError, PermissionError):
+        pass
```
It feels like this code might be neater having a single loop around common_license_names, instead of multiple. Not sure if this works, but something like...
```
dir_path = os.path.join(chromium_root, directory)
if not os.path.exists(dir_path):
    return

dirs_to_search = [dir_path]

try:
    first_level_dirs = [
        os.path.join(dir_path, d)
        for d in os.listdir(dir_path)
        if os.path.isdir(os.path.join(dir_path, d))
    ]
    dirs_to_search.extend(first_level_dirs)

    second_level_dirs = [
        os.path.join(subdir, d)
        for subdir in first_level_dirs
        for d in os.listdir(subdir)
        if os.path.isdir(os.path.join(subdir, d))
    ]
    dirs_to_search.extend(second_level_dirs)
except (OSError, PermissionError):
    pass

for license_name in common_license_names:
    for search_dir in dirs_to_search:
        license_path = os.path.join(search_dir, license_name)
        if os.path.isfile(license_path):
            rel_path = os.path.relpath(license_path, chromium_root)
            print(f"  Found fallback license file: {rel_path}")
            license_files_set.add(license_path)
```

caneraltinbasak replied:
I'll try

t-bashir-bs commented on meta-chromium/recipes-browser/chromium/generate_license_list.py:
```diff
@@ -86,22 +91,145 @@ def find_chromium_licenses(chromium_root):
+    # Walk through the entire chromium tree looking for license files
+    for root, dirs, files in os.walk(chromium_root):
+        # Skip some obviously non-relevant directories for performance
+        dirs[:] = [
```
Does this filter on dirs do anything, as I can't see it being used in the loop for anything and we always loop everything in files? Also, I think ".git" and ".svn" aren't needed as they will be excluded with ```not d.startswith(".")```

caneraltinbasak replied:
it's the list that `os.walk()` uses to determine which subdirectories to traverse next. But you are right, .git and .svn is not needed.

caneraltinbasak commented:
Now I think about it, I guess the licenses generated by the original version is better. Because Chromium relies on many thirdparty tools, which doesn't get shipped as part of the product. We might be listing them all unnecessarily if we use the comprehensive mode. I've addressed the review comments. I'll close the pull request but keep the branch alive for future reference.
